File: Find Bottom Left Tree Value/Find_Bottom_Left_Tree_Value.py
# Runtime: 56 ms, faster than 53.52% of Python3 online submissions for Find Bottom Left Tree Value.
# Memory Usage: 15.6 MB, less than 14.29% of Python3 online submissions for Find Bottom Left Tree Value.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def findBottomLeftValue(self, root: TreeNode) -> int:
        queue = []
        queue.append(root)        

        res = root.val
        while queue:
            size = len(queue) - 1
            root = queue.pop(0)

            res = root.val
            if root.left:
                queue.append(root.left)
            if root.right:
                queue.append(root.right)

            for _ in range(size):
                root = queue.pop(0)
                if root.left:
                    queue.append(root.left)
                if root.right:
                    queue.append(root.right)

        return res

File: Find Bottom Left Tree Value/Find_Bottom_Left_Tree_Value.cpp
// 基本思路 二叉树的广度优先搜索
// Runtime: 20 ms, faster than 99.31% of C++ online submissions for Find Bottom Left Tree Value.
// Memory Usage: 20.9 MB, less than 100.00% of C++ online submissions for Find Bottom Left Tree Value.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution 
{
public:
    int findBottomLeftValue(TreeNode* root) 
    {
        queue<TreeNode*> q;
        q.push(root);

        int res = root->val;
        while (!q.empty())
        {
            int size = q.size() - 1;
            root = q.front();
            q.pop();

            res = root->val;

            if (root->left) q.push(root->left);
            if (root->right) q.push(root->right);

            for (; size > 0; size--)
            {
                root = q.front();
                q.pop();

                if (root->left) q.push(root->left);
                if (root->right) q.push(root->right);
            }
        }
        return res;
    }
};